File: 653_Two_Sum_IV_-_Input_is_a_BST.java
/*
https://leetcode.com/problems/two-sum-iv-input-is-a-bst/
Given a Binary Search Tree and a target number, return true if there exist two elements in the BST such that their
sum is equal to the given target.

Sets stores seen values. Fro each node, check if complimnet has been seen, if not add to set and recurse on thildren.
The fact that it is a BST rather than a binary tree is irrelevant.
Time - O(n)
Space - O(n)
*/

class Solution {

    public boolean findTarget(TreeNode root, int k) {

        Set seen = new HashSet<Integer>();
        return helper(root, k, seen);
    }

    private boolean helper(TreeNode node, int k, Set seen) {

            if (node == null)
                return false;
            if (seen.contains(k - node.val))
                return true;

            seen.add(node.val);
            return (helper(node.left, k, seen) || helper(node.right, k, seen));
    }

}

File: 653_Two_Sum_IV_-_Input_is_a_BST.py
_author_ = 'jake'
_project_ = 'leetcode'

# https://leetcode.com/problems/two-sum-iv-input-is-a-bst/
# Given a Binary Search Tree and a target number, return true if there exist two elements in the BST such that their
# sum is equal to the given target.

# Traverse the BST. Preorder is used but could be postorder or inorder. For each node, check if k - node.val has
# already been visited. If node, add noe.val to the visited set. Would work equally well for a binary tree that is not
# a BST.
# Time - O(n)
# Space - O(n)

class Solution(object):
    def findTarget(self, root, k):
        """
        :type root: TreeNode
        :type k: int
        :rtype: bool
        """
        visited = set()

        def traverse(node):

            if not node:
                return False

            if k - node.val in visited:
                return True
            visited.add(node.val)

            return traverse(node.left) or traverse(node.right)

        return traverse(root)